Get access list share by email recipients

ASerbinski opened the pull request:
Close https://github.com/nextcloud/server/issues/32629

## Summary: 

Return the recipients for an email share. 

We already have `mail`, but it's unused / broken.

The backstory: 

In https://github.com/nextcloud/server/pull/2834 `mail => bool` was documented but not implemented. 

![image](https://github.com/nextcloud/server/assets/3902676/218004c3-3fb8-4900-9f02-5175c52a7853)


Commit to change mail to public: https://github.com/nextcloud/server/pull/2834/commits/cab41118f67e6a4743bac2cfcfa69a4bee28f2a3

ASerbinski committed:
ShareByMailProvider: getAccessList: return full list of recipients

Previously was returning only boolean true if the Node was shared
by email, or false if not. Now provides an array containing the
list of email share recipients.

Signed-off-by: Adam Serbinski <adam@serbinski.com>

ASerbinski committed:
ShareByMailProvider: getAccessList: add token to returned array

This allows the share URI to be regenerated.

Signed-off-by: Adam Serbinski <adam@serbinski.com>

CarlSchwan commented on apps/sharebymail/lib/ShareByMailProvider.php:
```diff
@@ -1194,21 +1194,30 @@ public function getAccessList($nodes, $currentAccess) {
 		$cursor = $qb->executeQuery();
 
-		$mail = $cursor->fetch() !== false;
+		$mail = array();
```
```suggestion
		$mail = [];
```

ASerbinski replied:
Suggestion pushed

CarlSchwan commented on lib/private/Share20/Manager.php:
```diff
@@ -1668,15 +1669,18 @@ public function userDeletedFromGroup($uid, $gid) {
 	 *  public => bool
-	 *  mail => bool
+	 *  mail => [
+         *      'email1@maildomain1' => ['node_id' => 42, 'token' => 'aBcDeFg'],
+         *      'email2@maildomain2' => ['node_id' => 23, 'token' => 'hIjKlMn'],
+         *  ]
```
```suggestion
		 *      'email1@maildomain1' => ['node_id' => 42, 'token' => 'aBcDeFg'],
		 *      'email2@maildomain2' => ['node_id' => 23, 'token' => 'hIjKlMn'],
		 *  ]
```

ASerbinski replied:
Suggestion pushed

ASerbinski commented on lib/public/Share/IManager.php:
```diff
@@ -272,15 +273,17 @@ public function userDeletedFromGroup($uid, $gid);
 	 *  remote => bool,
 	 *  public => bool
-	 *  mail => bool
+	 *  mail => ['email1@maildomain1', 'email2@maildomain2']
```
Yes, it certainly is an API break. I went through the history and it never actually used the correct array key.

ASerbinski committed:
ShareByMailProvider: getAccessList: recommended style changes

Signed-off-by: Adam Serbinski <adam@serbinski.com>

kesselb commented:
Hi @ASerbinski 

Thanks for your pull request :+1: 

Sorry for not getting back to you sooner. 
Would you mind rebasing your branch to pull the latest changes?

kesselb commented:
Would you mind splitting the bug fix and enhancement into own pull requests? 

As Carl mentioned, the enhancement comes with an API break. 
That's harder to review and accept than a bug report.

ASerbinski commented:
Hi @kesselb ;

As you've requested, I've rebased on master and split the bug fix into https://github.com/nextcloud/server/pull/36305

This PR now depends on that bug fix.

ASerbinski committed:
ShareByMailProvider: getAccessList: set both array keys 'public' and 'mail'

To preserve prior behaviour where 'public' was set 'true' if there
are any mail recipients. The 'mail' key will be an array of email
recipients.

Signed-off-by: Adam Serbinski <adam@serbinski.com>

blizzz commented:
Failing test seem related, for example:

```
There was 1 failure:

1) OCA\ShareByMail\Tests\ShareByMailProviderTest::testGetAccessList
Failed asserting that true is false.

/home/runner/actions-runner/_work/server/server/apps/sharebymail/tests/ShareByMailProviderTest.php:1093
```

AndyScherzinger commented:
@ASerbinski see comment from blizzz https://github.com/nextcloud/server/pull/32631#issuecomment-2010752483

ASerbinski committed:
ShareByMailProvider: getAccessList: set correct value for 'public'

PDOStatement::rowCount behavior is undefined for SELECT statements
for some database types, therefore manually set the value for 'public'
based on actual results fetched.

Signed-off-by: Adam Serbinski <adam@serbinski.com>

ASerbinski commented:
@AndyScherzinger @blizzz ;

Added a commit based on php documentation; https://www.php.net/manual/en/pdostatement.rowcount.php

"For statements that produce result sets, such as SELECT, the behavior is undefined"

Passing applicable unit test;

```
$ ./autotest.sh sqlite apps/sharebymail/tests/ShareByMailProviderTest.php
...
Time: 00:01.135, Memory: 36.00 MB

OK (37 tests, 186 assertions)
```


Other failures are unrelated and seem to refer to some kind of network failure.

ASerbinski commented:
@AndyScherzinger is it good now?

AndyScherzinger commented:
I am AFK but asked @sorbaugh to lend a hand in getting it merged

AndyScherzinger commented:
@ASerbinski merged 👍 Fancy writing a piece for the documentation repo to make the command easy to discover for others? 🙏

Thanks for the work on implementing this feature and for yet another contribution 🎉💯🚀

ASerbinski commented:
Thank you very much!
I'll give the documentation a shot.
